pkg/disk/filesystem.go

```diff
@@ -9,47 +9,106 @@ import (
 	"github.com/google/uuid"
 )
 
-type MkfsOption int
+type MkfsOption interface {
+	json.Marshaler
+	isMkfsOption()
+	Clone() MkfsOption
+}
 
-const ( // MkfsOption type enum
-	MkfsVerity MkfsOption = iota // Enable fs-verity option if needed (typically for EXT4)
-)
+type MkfsOptionVerity struct{}
 
-func getMkfsOptionMapping() []string {
-	return []string{"verity"}
+func (MkfsOptionVerity) isMkfsOption()     {}
+func (MkfsOptionVerity) Clone() MkfsOption { return &MkfsOptionVerity{} }
+func (o MkfsOptionVerity) MarshalJSON() ([]byte, error) {
+	return json.Marshal("verity")
 }
 
-// String converts MkfsOption into a human readable string
-func (option MkfsOption) String() string {
-	return getMkfsOptionMapping()[int(option)]
+type MkfsOptionGeometry struct {
+	Heads           int `json:"heads" yaml:"heads"`
+	SectorsPerTrack int `json:"sectors-per-track" yaml:"sectors-per-track"`
 }
 
-func unmarshalHelper(data []byte, mapping []string) (int, error) {
-	var stringInput string
-	err := json.Unmarshal(data, &stringInput)
-	if err != nil {
-		return 0, err
-	}
-	for n, str := range mapping {
-		if str == stringInput {
-			return n, nil
-		}
+func (MkfsOptionGeometry) isMkfsOption() {}
+func (o MkfsOptionGeometry) Clone() MkfsOption {
+	return &MkfsOptionGeometry{Heads: o.Heads, SectorsPerTrack: o.SectorsPerTrack}
+}
+func (o MkfsOptionGeometry) MarshalJSON() ([]byte, error) {
+	// Use an alias to avoid infinite recursion in MarshalJSON
+	type geomAlias MkfsOptionGeometry
+	type withType struct {
+		Type string `json:"type"`
+		geomAlias
 	}
-	return 0, fmt.Errorf("invalid mkfsoption: %s", stringInput)
+	return json.Marshal(withType{
+		Type:      "geometry",
+		geomAlias: geomAlias(o),
+	})
 }
 
-// UnmarshalJSON converts a JSON string into an MkfsOption
-func (option *MkfsOption) UnmarshalJSON(data []byte) error {
-	val, err := unmarshalHelper(data, getMkfsOptionMapping())
-	if err != nil {
-		return err
+type optionFactory func() MkfsOption
+
+var registry = map[string]optionFactory{
+	"verity":   func() MkfsOption { return &MkfsOptionVerity{} },
```

Comment:
Ideally we would (intendant of the implementation via interface or flat struct) some sort of validation that checks that the given filesystem type supports the options so that we error early if someone tries to enable "verity" on vfat by accident.

Reply:
done